File: verification-v194.py
#!/usr/bin/env python3
from pathlib import Path
import json, re, subprocess, sys, zipfile
ROOT=Path(__file__).resolve().parent
checks=[]
def check(name, cond, detail=''):
    checks.append((name,bool(cond),str(detail)))
    print(('PASS' if cond else 'FAIL')+f' | {name}'+(f' | {detail}' if detail else ''))
    return bool(cond)

def text(name): return (ROOT/name).read_text(encoding='utf-8',errors='replace')
core=text('core.js'); demo=text('demo-data.js'); services=text('services.js'); app=text('app.js'); styles=text('styles.css'); index=text('index.html'); sw=text('service-worker.js')

check('Core version is REV 1.0.94', "ProtoLab.VERSION = '1.0.94-poc'" in core)
check('HTML badge is REV 1.0.94', 'REV 1.0.94' in index)
for f in ['styles.css','core.js','demo-data.js','repository.js','services.js','app.js']:
    check(f'Cache bust updated: {f}', f'{f}?v=1.0.94' in index)
check('Service-worker cache reset is 1.0.94', '1.0.94' in sw)

for f in ['core.js','demo-data.js','repository.js','services.js','app.js','service-worker.js']:
    p=subprocess.run(['node','--check',str(ROOT/f)],capture_output=True,text=True)
    check(f'JavaScript syntax: {f}',p.returncode==0,(p.stderr or '').strip())

check('ISO-week formatter exists', 'function isoWeekPartsV1094' in app)
check('Grouped live calendar-axis formatter exists', 'function v1094GroupedCalendarAxis' in app and 'CW ${String(p.week).padStart' in app)
check('Month, ISO week-year and day layers styled', all(s in styles for s in ['swim-month-group-v1094','swim-week-group-v1094','swim-weekday-v1094']))
check('Total days late filter tile is present', "'Total days late'" in app and "tile('lateDays'" in app)
check('All commitment-health tile is present', "tile('all','All'" in app)
check('Commitment movement tile is a filter', "tile('movement','Commitment movement'" in app and 'data-v1094-health-filter' in app)
check('Build commitment health refresh honors filters', 'function v1094RefreshHealthCard' in app and 'filteredPlanningHealthV1094' in app)
check('Detailed bookings rendered section removed', '<summary>Detailed bookings and estimate basis</summary>' not in app)
check('Tiered AUTO PLAN control is active', 'AUTO PLAN · TIERED' in app and 'v1094OpenTieredPlanner' in app)
check('Escalation Mode control is active', 'ESCALATION MODE' in app and 'v1094RunEscalation' in app)

for token in ['planningPortfolioScoreV1094','planningWorsenedProjectsV1094','planningImprovementSuggestionsV1094','buildAutoPlanTiersV1094','buildEscalationPlanV1094','planPortfolioCandidateV1094']:
    check(f'Planner function exists: {token}', token in services)
check('Green policy can forbid recovery', "allowTraining:false" in services and "allowReadiness:false" in services)
check('Recovery-required path is explicit', "e.code='RECOVERY_REQUIRED'" in services)
check('Capacity failure is bounded and recoverable', "e.code='CAPACITY_UNRESOLVABLE'" in services and 'planningHorizonEnd' in services)
check('Legacy multi-year raw horizon message removed', 'no conflict-free slot exists within the controlled planning horizon' not in services+app)
check('UI explicitly refuses remote multi-year best-feasible output', 'rather than presenting a remote multi-year date as “best feasible”' in app)
check('Red tier requires whole-lab objective improvement', 'red.metrics.score<referenceScore' in services)
check('Red tier trade-off requires actual worsened project', 'red.worsened.length>0' in services)
check('Newly planned request is not falsely classified as worsened project', 'A newly planned request that was previously unplanned is not a project that became later' in services)
check('Escalation is target-first', 'if(targetRequestId){const i=rows.findIndex' in services and 'rows.unshift(rows.splice(i,1)[0])' in services)
check('Escalation evaluates multiple portfolio orderings', "['priorityDue','duePriority','commitment']" in services)
check('Planner candidate runs final invariant audit', 'planningIntegrityAudit' in services and 'validateInvariants' in services)
check('Capability-improvement suggestions include training', 'planningImprovementSuggestionsV1094' in services and re.search(r'train|training|qualification',services,re.I))

check('Provider-neutral team identity source exists', 'identityProvider' in core and 'externalGroupKey' in core)
check('Provider-neutral principal identity key exists', 'externalPrincipalId' in core)
check('Project-team role bindings exist', 'roleBindings' in core)
check('Project-team approval rights exist', 'APPROVAL_RIGHT_OPTIONS' in core and 'approvalRights' in core)
check('Project-team approval enforcement helper exists', 'projectTeamUserCanApprove' in core)
check('Project-team assignee resolution helper exists', 'resolveProjectTeamAssignee' in core)
check('Project team UI exposes roles and approval rights', 'Project Teams, Roles & Approval Rights' in app and 'External principal key' in app)
check('Project team UI exposes identity source', 'Identity source' in app)

check('Demo future-project probabilities use fractions', all(x in demo for x in ['probability:0.70','probability:0.45','probability:0.60']))
check('Legacy whole-percent probability migration normalizes to fractions', 'if(v>1&&v<=100)v/=100' in core)

# Retention checks from prior controlled workflow revisions.
check('REV 1.0.93 control-plan approval invalidation retained', 'invalidateControlPlanApproval' in app)
check('Control-plan approval path UI retained', 'v1093-cp-path' in app or 'cp-path' in app)
check('Final closeout is an explicit planned booking', 'Final quality review / release / handover' in services)

browser_path=ROOT/'VERIFICATION_BROWSER_v1.0.94.json'
check('Browser verification JSON exists',browser_path.exists())
if browser_path.exists():
    b=json.loads(browser_path.read_text())
    check('Browser: ISO year boundary correct', b.get('isoBoundary')=={'year':2026,'week':53}, b.get('isoBoundary'))
    check('Browser: Detailed bookings absent', b.get('hasDetails') is False)
    check('Browser: pipeline probabilities normalized', float(b.get('maxPipelineProbability',99))<=1, b.get('maxPipelineProbability'))
    check('Browser: All filter tile rendered', bool(b.get('allTile')))
    check('Browser: Total days late tile rendered', bool(b.get('totalLateTile')))
    check('Browser: grouped month axis rendered', len(b.get('calendarMonths') or [])>=2, b.get('calendarMonths'))
    check('Browser: ISO week-year axis rendered', all('CW ' in x and '· 2026' in x for x in (b.get('calendarWeeks') or [])[:2]), (b.get('calendarWeeks') or [])[:2])
    check('Browser: Commitment movement tile filters', b.get('activeFilter')=='movement' and bool(b.get('movementActiveClass')))
    check('Browser: tier cards rendered', int(b.get('tierCards') or 0)>=3,b.get('tierCards'))
    check('Browser: no raw horizon leak', b.get('rawHorizonLeak') is False)
    check('Browser: no multi-year best-feasible leak', b.get('multiYearBestLeak') is False)
    check('Browser: Project Team authority positive/negative check', bool((b.get('teamAuthority') or {}).get('ok')), b.get('teamAuthority'))
    eng=b.get('engine') or {}
    check('Engine: tier analysis returned recommendation', eng.get('recommended') in {'green','yellow','red'}, eng.get('recommended'))
    check('Engine: Red gate passed only under rule', eng.get('redRule') is True,eng.get('redRule'))
    check('Engine: no forecast beyond near-term test year', int(eng.get('maxYear') or 9999)<=2027,eng.get('maxYear'))
    hg=b.get('horizonGuard') or {}
    check('Adversarial horizon guard returns no fake recommendation', hg.get('recommended') is None and not hg.get('eligible'),hg)
    check('Adversarial horizon guard does not mutate live plan', hg.get('liveBookingsUnchanged') is True)
    check('Adversarial horizon guard does not leak raw legacy message', hg.get('rawLeak') is False)
    ex=b.get('escalationEngine') or {}
    check('Escalation produces valid target-first alternative', ex.get('ok') is True and ex.get('targetAfter') and ex.get('targetBefore'),ex)
    check('Escalation result has zero invariants/collisions/readiness failures', all(int(ex.get(k,99))==0 for k in ['invariants','collisions','readiness']),ex)
    check('Browser page completed without JS page errors', not b.get('pageErrors'),b.get('pageErrors'))

shot=ROOT/'REV1094_MASTER_PLANNER_PROOF.png'
check('Master Planner visual proof exists',shot.exists() and shot.stat().st_size>10000, shot.stat().st_size if shot.exists() else 0)

passed=sum(1 for _,ok,_ in checks if ok); total=len(checks); failed=total-passed
out=ROOT/'VERIFICATION_RESULTS_v1.0.94.txt'
out.write_text('\n'.join([f'LabOS REV 1.0.94 static + browser verification',f'Passed: {passed}',f'Failed: {failed}',f'Total: {total}','']+[f'{"PASS" if ok else "FAIL"} | {name}'+(f' | {detail}' if detail else '') for name,ok,detail in checks])+'\n',encoding='utf-8')
print(f'\nSUMMARY {passed}/{total} passed; {failed} failed')
sys.exit(0 if failed==0 else 1)

File: verification-v194.js
const fs=require('fs');
global.window=global;global.crypto=require('crypto').webcrypto;
const root=__dirname,read=f=>fs.readFileSync(`${root}/${f}`,'utf8');
for(const f of ['core.js','demo-data.js','services.js','repository.js'])eval(read(f));
const P=global.ProtoLab,app=read('app.js'),services=read('services.js'),index=read('index.html'),css=read('styles.css');
const checks=[];
function check(name,ok,detail=''){checks.push({name,ok:!!ok,detail});console.log(`${ok?'PASS':'FAIL'} | ${name}${detail?` | ${detail}`:''}`)}
check('Version and schema',P.VERSION==='1.0.94-poc'&&P.SCHEMA_VERSION===33,`${P.VERSION} / ${P.SCHEMA_VERSION}`);
check('Cache-busted release assets',(index.match(/1\.0\.94-r6/g)||[]).length>=6);
for(const f of ['core.js','demo-data.js','repository.js','services.js','app.js','service-worker.js']){const r=require('child_process').spawnSync(process.execPath,['--check',`${root}/${f}`],{encoding:'utf8'});check(`JavaScript syntax · ${f}`,r.status===0,(r.stderr||'').trim().slice(0,160));}
const fresh=P.createDemoState();P.ensureEnterpriseModel(fresh);
check('Fresh state invariants',P.validateInvariants(fresh).length===0,JSON.stringify(P.validateInvariants(fresh).slice(0,3)));
check('Pipeline probability normalized',fresh.pipelineProjects.every(x=>x.probability>=0&&x.probability<=1),fresh.pipelineProjects.map(x=>x.probability).join(','));
const old=P.deepClone(fresh);old.schemaVersion=31;const migrated=P.MigrationService.migrate(old);
check('Schema 31 → 33 migration',migrated.schemaVersion===33&&P.validateInvariants(migrated).length===0,`schema ${migrated.schemaVersion}`);
check('Project-team assignments migrated',migrated.requestingTeams.some(x=>x.assignments?.length));
const team=migrated.requestingTeams.find(x=>x.assignments?.length),assignment=team?.assignments?.[0];if(assignment){assignment.identity={provider:'entra-id',subject:'future-object-id',email:'future@example.test'};check('Provider-neutral principal resolution',P.projectTeamCan(migrated,team.id,{provider:'entra-id',subject:'future-object-id'},assignment.approvalRights[0]));}
check('Grouped month/week/day calendar',app.includes('swim-calendar-month-v1094')&&app.includes('swim-calendar-week-v1094')&&app.includes('v1094IsoWeekYear')&&css.includes('.swim-calendar-day-v1094'));
check('Total days-late KPI and health filters',app.includes('data-v1094-total-late')&&app.includes('data-v1094-health-filter')&&app.includes("['all','All']"));
check('Detailed booking section removed from rendered planner',app.includes("page.querySelectorAll('details.planning-details').forEach")&&app.includes("closest('.card')?.remove()"));
check('Yellow option requires explicit acceptance',app.includes('data-v1094-approve-yellow')&&app.includes('Yellow is an active planning decision'));
check('Tiered score guard exists',app.includes('NO_PORTFOLIO_IMPROVEMENT')&&app.includes('after.objective>=before.objective'));
check('Escalation replans target then portfolio',app.includes('v1094RunEscalation')&&app.includes('v1076TargetFirstPortfolio(proposed,id)'));
check('Capacity errors are decision-safe',!services.includes('no conflict-free slot exists')&&!services.includes('multi-year future date')&&services.includes('controlled capacity-recovery decision'));
check('Bounded interval-jump search retained',services.includes('planningHorizonEnd')&&services.includes('blockingEnds')&&services.includes('nextSearchPoint'));
check('One controlled planning-task manifest is shared',typeof P.planningTaskManifest==='function'&&typeof P.planningTaskDrift==='function'&&app.includes('v1094ControlledPlanningTasks')&&app.includes('v1094AlignWorkspacePlanningTasks'));
const singles=[];for(const r of fresh.requests.filter(x=>!['DELIVERED','CLOSED','RELEASED'].includes(x.status)).slice(0,8)){const s=P.deepClone(fresh),t=Date.now();try{const changes=new P.PlannerService().autoPlan(s,r.id),audit=P.planningIntegrityAudit(s);singles.push({id:r.id,ok:audit.ok,changes:changes.length,ms:Date.now()-t})}catch(e){singles.push({id:r.id,ok:false,code:e.code,ms:Date.now()-t})}}
check('Eight representative single plans are valid',singles.every(x=>x.ok),JSON.stringify(singles));
check('Single-plan performance bounded',singles.every(x=>x.ms<2000),`${Math.max(...singles.map(x=>x.ms))} ms max`);
const aligned=P.deepClone(fresh),alignedRequest=aligned.requests.find(x=>!['DELIVERED','CLOSED','RELEASED'].includes(x.status));new P.PlannerService().autoPlan(aligned,alignedRequest.id);const manifest=P.planningTaskManifest(aligned,alignedRequest.id),planned=aligned.bookings.filter(x=>x.requestId===alignedRequest.id&&!/complete|actual|done/i.test(String(x.status||'')));
check('AUTO-PLAN exactly matches controlled task manifest',P.planningTaskDrift(aligned,alignedRequest.id).length===0&&manifest.length===planned.length,`${manifest.length} controlled tasks / ${planned.length} bookings`);
check('Manifest covers process, tests and final handover',manifest.some(x=>x.kind==='process')&&manifest.some(x=>x.kind==='test')&&manifest.at(-1)?.kind==='closeout',manifest.map(x=>`${x.order}:${x.kind}`).join(','));
const renamed=P.deepClone(aligned),renamedBooking=renamed.bookings.find(x=>x.requestId===alignedRequest.id&&x.taskType==='test');if(renamedBooking)renamedBooking.stepName='Wrong sticky-menu test';check('Fault injection detects a mislabelled test',!!renamedBooking&&P.planningTaskDrift(renamed,alignedRequest.id).some(x=>/label/.test(x)));
const missing=P.deepClone(aligned),missingTest=missing.bookings.find(x=>x.requestId===alignedRequest.id&&x.taskType==='test');if(missingTest)missing.bookings=missing.bookings.filter(x=>x.id!==missingTest.id);check('Fault injection detects a missing planned test',!!missingTest&&P.planningTaskDrift(missing,alignedRequest.id).some(x=>/Missing planned task/.test(x)));
const duplicate=P.deepClone(aligned),dupSource=duplicate.bookings.find(x=>x.requestId===alignedRequest.id&&x.taskType==='test');if(dupSource)duplicate.bookings.push({...P.deepClone(dupSource),id:'FI-DUPLICATE-TEST'});check('Fault injection detects a duplicate planned test',!!dupSource&&P.planningTaskDrift(duplicate,alignedRequest.id).some(x=>/Duplicate planned task/.test(x)));
const portfolio=P.deepClone(fresh),pt=Date.now(),results=new P.PlannerService().autoPlanPortfolio(portfolio),portfolioMs=Date.now()-pt,failed=results.filter(x=>!x.ok),portfolioAudit=P.planningIntegrityAudit(portfolio);
check('Full portfolio plans atomically',!failed.length&&results.atomicApplied===true,`${results.length} builds / ${failed.length} failed`);
check('Full portfolio passes integrity audit',portfolioAudit.ok,JSON.stringify(portfolioAudit.counts));
check('Portfolio performance bounded',portfolioMs<10000,`${portfolioMs} ms`);
const blocked=P.deepClone(fresh),target=blocked.requests.find(r=>!['DELIVERED','CLOSED','RELEASED'].includes(r.status));blocked.settings.planningHorizonDays=7;blocked.settings.planningRecoveryDays=0;blocked.planningEvents=[{id:'FI-CLOSE',type:'Lab closure',scope:'lab',start:'2025-01-01T00:00:00.000Z',end:'2029-12-31T23:59:59.000Z',active:true,reason:'fault injection'}];const before=JSON.stringify(blocked),bt=Date.now();let capacityError=null;try{new P.PlannerService().autoPlan(blocked,target.id)}catch(e){capacityError=e}
check('Capacity exhaustion returns structured blocker',capacityError?.code==='CAPACITY_UNRESOLVABLE',capacityError?.message||'no error');
check('Capacity failure is transactional',before===JSON.stringify(blocked));
check('Capacity failure is fast',Date.now()-bt<1000,`${Date.now()-bt} ms`);
check('Capacity failure does not leak raw solver output',!/no conflict-free slot|multi-year future date|best feasible/i.test(capacityError?.message||''));
const noEquipment=P.deepClone(fresh);noEquipment.equipment.forEach(e=>e.status='Retired');const neBefore=JSON.stringify(noEquipment);let equipmentError=null;try{new P.PlannerService().autoPlan(noEquipment,target.id)}catch(e){equipmentError=e}
check('Unavailable equipment is rejected safely',!!equipmentError&&['ZERO_EQUIPMENT_CAPABILITY','CAPACITY_UNRESOLVABLE'].includes(equipmentError.code),equipmentError?.code||'no error');
check('Unavailable-equipment fault rolls back',neBefore===JSON.stringify(noEquipment));
const collision=P.deepClone(fresh),booking=collision.bookings.find(x=>x.equipmentId||x.staffId);collision.bookings.push({...P.deepClone(booking),id:'FI-DUPLICATE',stepId:'FI-DUPLICATE'});const collisionAudit=P.planningIntegrityAudit(collision);
check('Injected resource collision detected',!collisionAudit.ok&&(collisionAudit.equipmentOverlaps.length+collisionAudit.staffOverlaps.length)>0,JSON.stringify(collisionAudit.counts));
const invalidTeam=P.deepClone(fresh),it=invalidTeam.requestingTeams[0];it.assignments=[{id:'BAD',userId:'U-X',identity:{provider:'local',subject:'U-X'},roleId:'not-a-role',approvalRights:['not-a-right'],active:true}];const inv=P.validateInvariants(invalidTeam);
check('Invalid team role/right detected',inv.some(x=>/unknown role/.test(x))&&inv.some(x=>/unknown approval right/.test(x)),JSON.stringify(inv));
const failedChecks=checks.filter(x=>!x.ok);console.log(`\nSUMMARY | ${checks.length-failedChecks.length}/${checks.length} passed | ${failedChecks.length} failed`);process.exitCode=failedChecks.length?1:0;
